### PR TITLE
[ModelRunner V2] Misc minor simplifications and optimizations

### PR DESCRIPTION
**Pull request by njhill**

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/logprob.py [hunk 1]
```diff
@@ -73,19 +73,12 @@ def _ranks_kernel(
...
+    batch_size, vocab_size = logits.shape
     token_ids = token_ids.to(torch.int64)
     num_logprobs = token_ids.shape[1]
-    logprobs = torch.empty(
-        batch_size,
-        num_logprobs,
-        dtype=torch.float32,
-        device=logits.device,
-    )
+    logprobs = logits.new_empty((batch_size, num_logprobs))
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/sample/logprob.py]: IIUC, this function can be used for prompt logprobs. In that case, logits are bf16, whereas we want logprobs to be fp32. Let's use fp32 for clarity.

File: vllm/v1/worker/gpu/spec_decode/eagle_cudagraph.py [hunk 2]
```diff
@@ -35,16 +31,10 @@ def __init__(
...
-            cudagraph_mode = self.compilation_config.cudagraph_mode
-            if cudagraph_mode == CUDAGraphMode.FULL:
-                # NOTE(woosuk): For Eagle, we only use CUDA graphs for decode.
-                cudagraph_mode = CUDAGraphMode.FULL_DECODE_ONLY
-
-        self.cudagraph_mode = cudagraph_mode
+        self.cudagraph_mode = self.compilation_config.cudagraph_mode
+        if self.cudagraph_mode == CUDAGraphMode.FULL:
+            # NOTE(woosuk): For Eagle, we only use CUDA graphs for decode.
+            self.cudagraph_mode = CUDAGraphMode.FULL_DECODE_ONLY
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/spec_decode/eagle_cudagraph.py:37]: Can't `cudagraph_mode` be `None`? I thought we needed to cover that case.

File: vllm/v1/worker/gpu/cudagraph_utils.py [hunk 3]
```diff
@@ -41,11 +36,7 @@ def __init__(
         self.dp_size = vllm_config.parallel_config.data_parallel_size
         self.compilation_config = vllm_config.compilation_config
         assert self.compilation_config is not None
-        self.cudagraph_mode: CUDAGraphMode
-        if self.compilation_config.cudagraph_mode is None:
-            self.cudagraph_mode = CUDAGraphMode.NONE
-        else:
-            self.cudagraph_mode = self.compilation_config.cudagraph_mode
+        self.cudagraph_mode = self.compilation_config.cudagraph_mode
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/cudagraph_utils.py:39]: ditto

File: vllm/v1/worker/gpu/dp_utils.py [hunk 4]
```diff
@@ -13,26 +13,21 @@ def make_num_tokens_across_dp(dp_size: int, num_tokens: int) -> torch.Tensor | N
...
     group = get_dp_group().cpu_group
     tensor = torch.zeros(2, dp_size, dtype=torch.int32, device="cpu")
-    tensor[0][dp_rank] = num_tokens
-    tensor[1][dp_rank] = cudagraph_size
+    all_num_tokens, all_cudagraph_sizes = tensor
+    all_num_tokens[dp_rank] = num_tokens
+    all_cudagraph_sizes[dp_rank] = cudagraph_size
     dist.all_reduce(tensor, group=group)
-    return tensor[0], tensor[1]
+    return all_num_tokens, all_cudagraph_sizes
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/dp_utils.py]: I'm not sure whether this looks simpler. To me personally, explicit indexing (left) is easier to understand.

File: vllm/v1/worker/gpu/lora_utils.py [hunk 5]
```diff
@@ -41,7 +38,6 @@ def make_lora_inputs(
 
         active_lora_requests: set[LoRARequest] = set()
         for req_id in req_ids:
-            lora_request = self.lora_requests.get(req_id, None)
-            if lora_request is not None:
+            if (lora_request := self.lora_requests.get(req_id)) is not None:
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/lora_utils.py]: My personal preference is left. I personally never use this style 😅

File: vllm/v1/worker/gpu/model_runner.py [hunk 6]
```diff
@@ -474,28 +473,22 @@ def update_requests(self, scheduler_output: SchedulerOutput) -> None:
...
+        num_scheduled_tokens = np.fromiter(
+            map(req_num_tokens.get, req_ids), dtype=np.int32, count=num_reqs
         )
 
-        idx_mapping_list = [
-            self.req_states.req_id_to_index[req_id] for req_id in req_ids
-        ]
-        idx_mapping_np = np.array(idx_mapping_list, dtype=np.int32)
+        idx_mapping_iter = map(self.req_states.req_id_to_index.get, req_ids)
+        idx_mapping_np = np.fromiter(idx_mapping_iter, dtype=np.int32, count=num_reqs)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Maybe it’s just me, but I’m not very familiar with `np.fromiter`, so this version takes more mental effort to parse, while the left one doesn’t.

**Review by WoosukKwon — COMMENTED** [review]
Thanks for the PR! Left a couple comments on correctness, and a few on style.

File: vllm/v1/worker/gpu/spec_decode/eagle_cudagraph.py (hunk 2 above)
Comment (njhill) [reply vllm/v1/worker/gpu/spec_decode/eagle_cudagraph.py:37]: The type of this config field is `CUDAGraphMode` (not optional), and it's always set to a non-`None` value in `VllmConfig.__post_init__`. So it's reasonable here imo to assume non-`None` to keep this code simpler.

File: vllm/v1/worker/gpu/model_runner.py (hunk 6 above)
Comment (njhill) [reply vllm/v1/worker/gpu/model_runner.py]: As well as fewer loc, these changes are ~30% faster in microbenchmark and also reduce GC overhead since we materialize fewer batch-size python lists.

`np.fromiter` is just like 1D `np.array` except that it takes an iterable.. the name is kind of self-explanatory? :)

File: vllm/v1/worker/gpu/dp_utils.py (hunk 4 above)
Comment (njhill) [reply vllm/v1/worker/gpu/dp_utils.py]: Was mainly to avoid creating the same slices twice, agree it looks a bit messier though!

File: vllm/v1/worker/gpu/sample/logprob.py (hunk 1 above)
Comment (njhill) [reply vllm/v1/worker/gpu/sample/logprob.py]: Fixed

File: vllm/v1/worker/gpu/cudagraph_utils.py (hunk 3 above)
Comment (njhill) [reply vllm/v1/worker/gpu/cudagraph_utils.py:39]: See other response

**Review by WoosukKwon — APPROVED** [review]
LGTM
